core/src/banking_stage/transaction_scheduler/transaction_state_container.rs

```diff
@@ -0,0 +1,311 @@
+    std::collections::HashMap,
+};
+
+/// This structure will hold `TransactionState` for the entirety of a
+/// transaction's lifetime in the scheduler and BankingStage as a whole.
+///
+/// Transaction Lifetime:
+/// 1. Received from `SigVerify` by `BankingStage`
+/// 2. Inserted into `TransactionStateContainer` by `BankingStage`
+/// 3. Popped in priority-order by scheduler, and trnsitioned to `Pending` state
+/// 4. Processed by `ConsumeWorker`
+///   a. If consumed, remove `Pending` state from the `TransactionStateContainer`
+///   b. If retryable, transition back to `Unprocessed` state.
+///      Re-insert to the queue, and return to step 3.
+///
+/// The structure is composed of two main components:
+/// 1. A priority queue of wrapped `TransactionId`s, which are used to
+///    order transactions by priority for selection by the scheduler.
+/// 2. A map of `TransactionId` to `TransactionState`, which is used to
+///    track the state of each transaction.
+///
+/// When `Pending`, the associated `TransactionId` is not in the queue, but
+/// is still in the map.
+/// The entry in the map should exist before insertion into the queue, and be
+/// be removed only after the id is removed from the queue.
+///
+/// The container maintains a fixed capacity. If the queue is full when pushing
+/// a new transaction, the lowest priority transaction will be dropped.
+pub(crate) struct TransactionStateContainer {
+    priority_queue: MinMaxHeap<TransactionPriorityId>,
+    id_to_transaction_state: HashMap<TransactionId, TransactionState>,
+}
+
+impl TransactionStateContainer {
+    pub(crate) fn with_capacity(capacity: usize) -> Self {
+        Self {
+            priority_queue: MinMaxHeap::with_capacity(capacity),
+            id_to_transaction_state: HashMap::with_capacity(capacity),
+        }
+    }
+
+    /// Returns true if the queue is empty.
+    pub(crate) fn is_empty(&self) -> bool {
+        self.priority_queue.is_empty()
+    }
+
+    /// Returns the remaining capacity of the queue
+    pub(crate) fn remaining_queue_capacity(&self) -> usize {
+        self.priority_queue.capacity() - self.priority_queue.len()
+    }
+
+    /// Get an iterator of the top `n` transaction ids in the priority queue.
+    /// This will remove the ids from the queue, but not drain the remainder
+    /// of the queue.
+    pub(crate) fn take_top_n(
+        &mut self,
+        n: usize,
+    ) -> impl Iterator<Item = TransactionPriorityId> + '_ {
+        (0..n).map_while(|_| self.priority_queue.pop_max())
+    }
+
+    /// Serialize entire priority queue. `hold` indicates whether the priority queue should
+    /// be drained or not.
+    /// If `hold` is true, these ids should not be removed from the map while processing.
+    pub(crate) fn priority_ordered_ids(&mut self, hold: bool) -> Vec<TransactionPriorityId> {
+        let priority_queue = if hold {
+            self.priority_queue.clone()
```

Comment:
This would happen when we make the `ForwardAndHold` decision. In that case we don't want to drain our queue.

Unfortunately, heaps do not have an interface for iterating over these without draining.
Another option would have been to drain the iterator, then replace them if we have to hold.
Either way, because we replace the original priority queue, there's an allocation of O(n).